### ISSUE TITLE
Parser finds nonexistent tokens after unclosed delimiters

### ISSUE THREAD

**Issue by FabianWolff** [rust-lang/rust#86073]
Whenever there is an unclosed delimiter in the input, the parser tries to recover by inserting the expected closing delimiter somewhere. However, this leads to spurious error messages if another error occurs after recovering, e.g.:
```rust
#[i=i::
```
```
error: this file contains an unclosed delimiter
 --> test1.rs:1:8
  |
1 | #[i=i::
  |  -     ^
  |  |
  |  unclosed delimiter

error: expected identifier, found `]`
 --> test1.rs:1:8
  |
1 | #[i=i::
  |        ^ expected identifier

error: aborting due to 2 previous errors
```
or:
```rust
impl X { fn foo
```
```
error: this file contains an unclosed delimiter
 --> test2.rs:1:16
  |
1 | impl X { fn foo
  |        -       ^
  |        |
  |        unclosed delimiter

error: expected one of `(` or `<`, found `}`
 --> test2.rs:1:16
  |
1 | impl X { fn foo
  |        -       ^
  |        |       |
  |        |       expected one of `(` or `<`
  |        |       the item list ends here
  |        while parsing this item list starting here
```
i.e. the error messages say that certain tokens were found that clearly aren't present in the input. I can understand why this happens, but I think it is confusing, and that artificially inserted tokens should not appear in error messages like this.

**Comment by estebank** [comment]
This is somewhat a consequence of turning the TokenSequence into a TokenTree. I would love it if we could treat this better.

**Comment by estebank** [comment]
Current output:
```
error: this file contains an unclosed delimiter
 --> src/lib.rs:1:8
  |
1 | #[i=i::
  |  -     ^
  |  |
  |  unclosed delimiter
```
```
error: this file contains an unclosed delimiter
 --> src/lib.rs:1:16
  |
1 | impl X { fn foo
  |        -       ^
  |        |
  |        unclosed delimiter
```
